lib/streamlit/runtime/state/query_params.py

```diff
@@ -0,0 +1,126 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from dataclasses import dataclass, field
+from typing import Any, Dict, Iterator, List, MutableMapping, Union
+
+from streamlit.proto.ForwardMsg_pb2 import ForwardMsg
+
+
+@dataclass
+class QueryParams(MutableMapping[str, Any]):
+    """A lightweight wrapper of a dict that sends forwardMsgs when state changes.
+    It stores str keys with str and List[str] values.
+    """
+
+    _query_params: Dict[str, Union[List[str], str]] = field(default_factory=dict)
+
+    def __init__(self):
+        # avoid using ._query_params as that will use __setattr__,
+        # which itself relies on `_query_params` being defined
+        self.__dict__["_query_params"] = {}
+
+    def __iter__(self) -> Iterator[Any]:
+        return iter(self._query_params.keys())
```

Comment:
Worth checking if this is needed since this class already extends `MutableMapping`

Reply:
<img width="807" alt="Screenshot 2023-11-14 at 3 46 03 PM" src="https://github.com/streamlit/streamlit/assets/16749069/f3d52f46-6c10-4cda-b9d6-786a6aee9ff2">
The third column are abstract methods so yep seems necessary